components/builder-api-client/src/builder.rs

```diff
@@ -90,6 +92,24 @@ mod json_u64 {
     }
 }
 
+#[cfg(windows)]
+use crate::util::win_perm;
+
+/// Abstraction over platform-specific ways to model file permissions.
+enum Permissions {
+    /// Don't take any special action to set permissions beyond what
+    /// they are "normally" set to when they are created. Here,
+    /// "normal" denotes the low-level programming library sense,
+    /// rather than any particular domain-specific sense.
+    Standard,
+    /// Indicates that a file should be created with very specific
+    /// permissions.
+    #[cfg(windows)]
+    Explicit(Vec<win_perm::PermissionEntry>),
+    #[cfg(not(windows))]
+    Explicit(u32),
```

Comment:
Wonder if it is worth using [`std::fs::Permissions`](https://doc.rust-lang.org/stable/std/fs/struct.Permissions.html) here? It looks like it is really just a thin wrapper around a `u32` so maybe it does not buy us much. Probably fine either way.

Reply:
Yeah, I don't think that really buys us much at all; we'd still need some additional abstraction on top of that to deal with Windows stuff the way we would want to.

The main thing this type really does is provide a richer, permission-specific `Option` for us to use (as well as a facade behind which to hide all the platform-specific differences). I felt it was much better to have function signatures like 
```
do_something(..., Permissions::Standard, ...)
``` 

than

```
do_something(..., None, ...)
```

Reply:
I was not very clear. I definitely like the `Option` like wrapper. I was saying the enum could become:
```rust
pub enum Permissions {
    Standard,
    #[cfg(windows)]
    Explicit(Vec<win_perm::PermissionEntry>),
    #[cfg(not(windows))]
    Explicit(std::fs::Permissions),
}
```

Reply:
ah, I see what you meant :smile: